### ISSUE TITLE
Registers, temporary words?

### ISSUE THREAD

**Issue by wboeke** [CCurl/c3#17]
Am I missing something, or have the R0 etc and T0 etc been omitted in this version?
Btw: local variables for functions would be much more useful then these builtin variables, as they would be useable in recursive functions also.

W.Boeke

**Comment by CCurl** [comment]
For registers, it they HAVE to be lower case, so ... r0, not R0

For temporary words, they HAVE to ne UPPER case, so T0, not t0.

They are both in all version, including the current version.

**Comment by CCurl** [comment]
Note: the temporary words don't show up in the "WORDS" list, nor do the register operations, except for +regs and -regs, which DO show up in "WORDS". But they are there.

Sorry if that wasn't clear in the README. I'll update it to say that.

**Comment by wboeke** [comment]
If I try this:
: rot ( a b c--b c a ) +regs s3 s2 s1  r2 r3 r1  -regs ;
then c3 prints:
[+regs]??[s3]??[s2]??[s1]??[r2]??[r1]??-under-[r1]??[r3]??-under--under-[-regs]?? ok
In the previous version of c3 this worked well.
At that: a search for 'reg' in the .c file shows no result. 

And:
: sqrt ( n--0|sqrt ) dup 0 > if T0 else drop 0 then ;
causes a segfault. This also went well in the previous version.
I hope all this is my own stupid fault!

W.Boeke

**Comment by CCurl** [comment]
No ... it's me .... sorry ... the example in the README is old (and wrong now). It should be this:

\ The Babylon square root algorithm
: T0 ( n--sqrt ) dup 4 / begin >r dup r@ / r@ + 2 / dup r> - 0= until nip ;
: sqrt ( n--0|sqrt ) dup 0 > if T0 else drop 0 then ;


![image](https://github.com/CCurl/c3/assets/14334380/86ef6f10-6acf-4d40-9e2c-6280bf3a486f)

**Comment by CCurl** [comment]
Re local variables: since you have complete control over them, they can be used recursively if you want ... just do +regs and -regs in the function.

One thing ... the register stack is 100 deep by default, so it would support 10 recursions out of the box. If you want more, you can easily change it to be deeper.

I should make that a #define so it is one line of code to change for that ... it's currently hard-coded to 100, so there are 2 lines to change instead :).

**Comment by wboeke** [comment]
Unrelated:
Maybe it's a good idea to omit the 'inline' word in c3.c. Modern compilers will inline functions if needed, and at that, I get a lot of linker errors with options -O0 or -O1.
Another remark: in your 'make' script the option -O3 is used. Did you know  that everything is inlined then, yielding a big executable? At that, speed is not that important for a project like this.

I tried to simulate a recursive function with local variables, using those builtin registers. I did'nt find a streightforward solution, maybe you could give an example? 

W.Boeke

**Comment by CCurl** [comment]
In your example, you are never setting the value of any register. So r0 will always be 0.

Try this:

: fib ( n1--n2 ) +regs s0 r0 1 > if r0 1 - fib r0 2 - fib + else 1 then -regs ;

The registers are stored in an array ...

cell_t regs[REGS_SZ], regBase=0;

+regs simply provides does a "regBase += 10"

So you can see that the previous values of the registers are not kept.

Conversely, -regs simply does "regBase -= 10;
![image](https://github.com/CCurl/c3/assets/14334380/5b5cce47-99cd-45a5-bbf7-7cf72856b2e3)

**Comment by CCurl** [comment]
> Unrelated: Maybe it's a good idea to omit the 'inline' word in c3.c. Modern compilers will inline functions if needed, and at that, I get a lot of linker errors with options -O0 or -O1. Another remark: in your 'make' script the option -O3 is used. Did you know that everything is inlined then, yielding a big executable? At that, speed is not that important for a project like this.
> 
> W.Boeke
Good info, thanks. I am not familiar with the clang optimization settings. I will remove the "inline" from c3.c and see what it outputs.

I don't know if you use Windows/Visual Studio, but in that realm, there are "Debug" builds that are MUCH slower than "Release" builds. I was imagining that without the -O3, that it would produce something closer to the "Debug" build. That understanding is obviously incorrect.

What would you recommend for Linux? You know ... I am going for small (minimal), but not significantly slower because of missing optimizations.

**Comment by wboeke** [comment]
Thanks! The fibonacci ran okay. 10 fib was the maximum for a correct result.
I am not on Windows but on Linux (Fedora and Ubuntu).
Something like Visual Studio is not needed at all.
The compilers (gcc or clang) do what they should do, and not more. During development I use -O0: fast compilation and no automatic inlining, thus simple error messages during compilation.
I am investiging your c3, you must be a clever developer!
Regards,
W.Boeke

**Comment by CCurl** [comment]
Yes, 10 currently. But as I said, it is extremely easy to change if you need to support more.

**Comment by CCurl** [comment]
So I tried -O0 (fixing the errors by removing the "inline"). My 250MIL benches in (benches.c3) were much slower ... about twice as slow.

Then I tried -O1, and it was much better ... essentially the same times as the -O3 times.

Neither made much different in terms of the size of the executable (about 31K). The -O0 and -O1 versions actually were a little larger, not smaller.

I think I'll leave it at -O1.
